### PR TITLE
Infinite Spraycan Additions

### PR DESCRIPTION
**Pull request by querns**
## Infinite Spray Can (ISC)

This PR adds all the rest of the stuff I wanted to do with infinite spraycans but couldn't fit in before the feature freeze. None of these additions are crucial to the item's functionality, but they do vastly improve its usability. Merge it whenever you feel like.

### New Features

* **Locking:** The ISC can now be locked. Sneak-middle-click to toggle the lock on and off. A locked ISC can't change its color at all until you unlock it. Attempting to do so will print an error message in the player's HUD. There are sounds associated with locking and unlocking as well.
* **New item renderer:** The spraycan will now show a color swathe for its current color and whether or not it's locked. 
* **Color choice GUI:** Middle click the air to open the color choice GUI.
* **Pick color:** Middle click a block to copy the color of that block to the spraycan.
* **Ring of Loki support:** All spraycans (not just the ISC) now work with the Ring of Loki. Naturally, this also includes line mode.
* **Improved Tooltip:** The tooltip now hides all its help messages and won't show them until you press the SHIFT key.

### Screenshots

java_Z2rPj228xM

java_E37mSv63Bm

java_wd0VSfOeUr

### Videos

Loki Support:
Loki Support

The locking and unlocking sounds (sound on):
New Sounds

### CODE REVIEW

**Comment by S4mpsa** [comment]
I would argue this is an enhancement to an existing feature which was already added before the merge, and should be included even during feature freeze.

**Comment by Caedis** [comment]
That is fine then

File: src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java [hunk 1]
```diff
@@ -35,12 +45,21 @@ public byte getPacketID() {
...
+        aOut.writeBytes(name);
     }
 
     @Override
     public GTPacketNew decode(final ByteArrayDataInput aData) {
-        return new GTPacketInfiniteSpraycan(aData.readBoolean());
+        final int newColor = aData.readInt();
+        final int length = aData.readInt();
+        final byte[] name = new byte[length];
+        aData.readFully(name, 0, length);
```
Comment (Caedis) [inline src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java:60]: cant you use readUTF or writeUTF?

Comment (querns) [reply src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java:60]: `ByteBuf`s don't have a `writeUTF` method. I will try the `readUTF` method of the `ByteArrayDataInput` class and see if it works.

Comment (Caedis) [reply src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java:60]: All good, was just wondering

**Review by Caedis — APPROVED** [review]

File: src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java (hunk 1 above)
Comment (Caedis) [reply src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java:60]: Yeah i thought there was a writeString/readString and checked how other packets handled strings and saw it.

Comment (Lyfts) [reply src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java:60]: There's `ByteBufUtils.readUTF8String` and ofc a write equivalent

Comment (querns) [reply src/main/java/gregtech/api/net/GTPacketInfiniteSpraycan.java:60]: I tried those, and I kept getting end of stream exceptions.

**Pull request merged**